crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -1082,79 +1072,34 @@ impl ObjectFilter {
             };
         }
 
-        // Treat `object_ids` and `object_keys` as a single filter on IDs, and optionally versions,
-        // and compute the intersection of that.
-        let keys = intersect::field(self.keys(), other.keys(), |k, l| {
-            let mut combined = BTreeMap::new();
+        let object_ids = {
+            match (self.object_ids, other.object_ids) {
+                (Some(obj_ids), Some(other_obj_ids)) => {
+                    let set_a: BTreeSet<_> = obj_ids.iter().cloned().collect();
+                    let set_b: BTreeSet<_> = other_obj_ids.iter().cloned().collect();
+                    let intersection: Vec<_> = set_a.intersection(&set_b).cloned().collect();
+
+                    if intersection.is_empty() {
+                        return None;
+                    }
 
-            for (id, v) in k {
-                if let Some(w) = l.get(&id).copied() {
-                    combined.insert(id, intersect::field(v, w, intersect::by_eq)?);
+                    Some(intersection)
                 }
+                (Some(obj_ids), None) => intersect::field(Some(obj_ids), None, intersect::by_eq)?,
+                (None, Some(other_obj_ids)) => {
+                    intersect::field(Some(other_obj_ids), None, intersect::by_eq)?
+                }
+                _ => None,
             }
```

Comment:
This can be simplified to the following:

```rust
let object_ids = intersect::field(self.object_ids, other.object_ids, |a, b| {
    let a = BTreeSet::from_iter(a);
    let b = BTreeSet::from_iter(b);

    let intersection: Vec<_> = a.intersection(&b).cloned().collect();
    (!intersection.is_empty()).then_some(intersection)
});
```